File: src/main/java/com/floorestimatepro/persistence/ProjectStore.java
package com.floorestimatepro.persistence;
import java.util.List;
import java.util.ArrayList;
import com.floorestimatepro.model.Project;
import com.floorestimatepro.model.Room;
import com.floorestimatepro.model.PlanPoint;
import com.floorestimatepro.model.Obstacle;
import com.floorestimatepro.model.Calibration;
import com.google.gson.Gson;
import java.nio.file.Files;
import java.nio.file.Path;

record PointData(double x, double y){
    
}

record CalibrationData(double feetPerPixel){
}

record RoomData(List<PointData> listOfPoints){

}

record ObstacleData(List<PointData> listOfPoints){

}

record ProjectData(
    String imagePath,
    CalibrationData calibration,
    List<RoomData> rooms,
    List<ObstacleData> obstacles
){};


public class ProjectStore{
    public static ProjectData toData(Project project){

        CalibrationData calData = null;

        if(project.calibration() != null){
                calData = new CalibrationData(
                project.calibration().feetPerPixel()
            );
        }

        List<RoomData> roomDatas = new ArrayList<>();

        for(Room room: project.rooms()){
            List<PointData> pointDatas = new ArrayList<>();
                for(PlanPoint point: room.listOfPoints()){
                    pointDatas.add(new PointData(point.x(), point.y()));
                }
            roomDatas.add(new RoomData(pointDatas));
        }

       List<ObstacleData> obstacleDatas = new ArrayList<>();

        for(Obstacle obstacle: project.obstacles()){
            List<PointData> pointDatas = new ArrayList<>();
                for(PlanPoint point: obstacle.listOfPoints()){
                    pointDatas.add(new PointData(point.x(), point.y()));
                }
            obstacleDatas.add(new ObstacleData(pointDatas));
        }

        return new ProjectData(project.imagePath(), calData, roomDatas, obstacleDatas);
    }

    public static Project fromData(ProjectData data){
        Project project =  new Project();

        project.setImagePath(data.imagePath());

        if(data.calibration() != null){
            Calibration cal = new Calibration(data.calibration().feetPerPixel());
            project.setCalibration(cal);
        }
    
        for(RoomData roomData: data.rooms()){
            List<PlanPoint> points = new ArrayList<>();
            for(PointData pd : roomData.listOfPoints()){
                points.add(new PlanPoint(pd.x(), pd.y()));
            }
            project.addRoom(new Room(points));
        }

       for(ObstacleData obstacleData: data.obstacles()){
            List<PlanPoint> points = new ArrayList<>();
            for(PointData po : obstacleData.listOfPoints()){
                points.add(new PlanPoint(po.x(), po.y()));
            }
            project.addObstacle(new Obstacle(points));
        }

        return project;
    }

    public static void save(Project project, String filePath) throws Exception {
        ProjectData data = toData(project);

        String jsonText = new Gson().toJson(data);

        Files.writeString(Path.of(filePath), jsonText);
    }

    public static Project load(String filePath) throws Exception {
        String jsonText = Files.readString(Path.of(filePath));

        ProjectData data = new Gson().fromJson(jsonText, ProjectData.class);

        return fromData(data);
    }
}

File: python/src/floor_estimate_pro/persistence/project_store.py
from __future__ import annotations

import json
from pathlib import Path

from floor_estimate_pro.model.calibration import Calibration
from floor_estimate_pro.model.obstacle import Obstacle
from floor_estimate_pro.model.plan_point import PlanPoint
from floor_estimate_pro.model.project import Project
from floor_estimate_pro.model.room import Room


class ProjectStore:
    @staticmethod
    def to_data(project: Project) -> dict:
        cal_data = None
        if project.calibration() is not None:
            cal_data = {
                "feetPerPixel": project.calibration().feet_per_pixel
            }

        room_datas = []
        for room in project.rooms():
            point_datas = []
            for point in room.list_of_points:
                point_datas.append({"x": point.x, "y": point.y})
            room_datas.append({"listOfPoints": point_datas})

        obstacle_datas = []
        for obstacle in project.obstacles():
            point_datas = []
            for point in obstacle.list_of_points:
                point_datas.append({"x": point.x, "y": point.y})
            obstacle_datas.append({"listOfPoints": point_datas})

        return {
            "imagePath": project.image_path(),
            "calibration": cal_data,
            "rooms": room_datas,
            "obstacles": obstacle_datas,
        }

    @staticmethod
    def from_data(data: dict) -> Project:
        project = Project()
        project.set_image_path(data.get("imagePath"))

        cal = data.get("calibration")
        if cal is not None:
            project.set_calibration(Calibration(cal["feetPerPixel"]))

        for room_data in data.get("rooms") or []:
            points = []
            for pd in room_data.get("listOfPoints") or []:
                points.append(PlanPoint(pd["x"], pd["y"]))
            project.add_room(Room(points))

        for obstacle_data in data.get("obstacles") or []:
            points = []
            for po in obstacle_data.get("listOfPoints") or []:
                points.append(PlanPoint(po["x"], po["y"]))
            project.add_obstacle(Obstacle(points))

        return project

    @staticmethod
    def save(project: Project, file_path: str) -> None:
        data = ProjectStore.to_data(project)
        json_text = json.dumps(data)
        Path(file_path).write_text(json_text, encoding="utf-8")

    @staticmethod
    def load(file_path: str) -> Project:
        json_text = Path(file_path).read_text(encoding="utf-8")
        data = json.loads(json_text)
        return ProjectStore.from_data(data)
